### PR TITLE
MANTA-5139 MPU for mantav2 needs scalable GC

### PR DESCRIPTION
**Pull request by joshwilsdon**

### CODE REVIEW

File: lib/garbage-mpu-parts-cleaner.js [hunk 1]
```diff
@@ -0,0 +1,607 @@
...
+
+                self.addCounter('partDeleteBatchSeconds', elapsed);
+
+                if (err) {
+                    self.addCounter('partDeleteBatchErrors', 1);
+                    logLevel = 'error';
+                } else {
+                    assert.array(res.etags, 'res.etags');
+                    self.addCounter('partDeleteSuccessCount', res.etags.length);
+                    deletedParts = res.etags.length;
```
Comment (trentm) [inline lib/garbage-mpu-parts-cleaner.js:467]: XXX What does it mean if res.etags.length is less than deleteBatch.length?

File: package.json [hunk 2]
```diff
@@ -15,7 +15,7 @@
     "forkexec": "1.1.0",
     "jsprim": "2.0.0",
     "kthxbai": "0.4.0",
-    "moray": "3.5.0",
+    "moray": "git+https://github.com/joyent/node-moray.git#v4.0.0",
```
[the remark this answers is not in the document]
Comment (joshwilsdon) [reply package.json:18]: As you wish.

File: lib/garbage-mpu-cleaner.js [hunk 3]
```diff
@@ -0,0 +1,657 @@
...
+    // The default value here was chosen without data, it might be adjusted once
+    // we have data.
+    //
+    // mpu_batch_size sets the number of manta_upload records to read on each
+    // findObjects call (limit option in the moray request). The default value
+    // was chosen basically at random here since we have no real data or targets
+    // for the usage of this system. These might need to be adjusted once we
+    // have some production data for this system.
+    //
+    self.mpuCleanupAgeSecs = self.config.options.mpu_cleanup_age_seconds || 300; // XXX default 1h?
```
[the remark this answers is not in the document]
Comment (joshwilsdon) [reply lib/garbage-mpu-cleaner.js:111]: Ok! I guess if someone complains it's too short we can always adjust later.

File: lib/garbage-mpu-cleaner.js [hunk 4]
```diff
@@ -0,0 +1,657 @@
...
+
+                            //
+                            // We log the data received at `info` level if there
+                            // is any, so that we have an audit record of
+                            // everything we attempted to delete. If this ends
+                            // up being too much data, we can increase the
+                            // processing overhead and trim down the objects but
+                            // ideally we'd have at minimum:
+                            //
+                            // // XXX
```
[the remark this answers is not in the document]
Comment (joshwilsdon) [reply lib/garbage-mpu-cleaner.js:325]: Yes. Updated for the next push.

File: lib/garbage-mpu-cleaner.js [hunk 5]
```diff
@@ -0,0 +1,657 @@
...
+ * Copyright 2020 Joyent, Inc.
+ */
+
+//
+// This contains the code for GarbageMpuCleaner objects which are used by
+// the garbage-mpu-cleaner program to:
+//
+//   * pull MPU finalize records from Moray
+//   * list the parts for each upload using Electric Moray
+//   * delete the parts and directory using Electric Moray
```
[the remark this answers is not in the document]
Comment (joshwilsdon) [reply lib/garbage-mpu-cleaner.js:17]: The latter. There was never a design where each instance had its own emoray connection. I guess I'll update the comment to hopefully make this more clear.

File: lib/garbage-mpu-cleaner.js [hunk 6]
```diff
@@ -0,0 +1,657 @@
...
+                                    successfullyDeleted: ctx.successfullyDeleted
+                                },
+                                'Deleted manta_upload records.'
+                            );
+
+                            deletedRecords = ctx.successfullyDeleted.length;
+                            self.addCounter(
+                                'morayFinalizeDeleteCount',
+                                ctx.successfullyDeleted.length
+                            );
```
[the remark this answers is not in the document]
Comment (joshwilsdon) [reply lib/garbage-mpu-cleaner.js:483]: There was a bug here, yep. Nice catch.

File: lib/garbage-mpu-parts-cleaner.js [hunk 7]
```diff
@@ -0,0 +1,607 @@
...
+    // The actual limit for the Manta v2 MPU is 800 (MAX_PARTS) parts, but we
+    // set the limit to MAX_PARTS + 1 here so that we can blow up if there were
+    // too many. Because in that case we might not be collecting correctly.
+    var findOpts = {limit: MAX_PARTS + 1};
+    var listBegin;
+    var record = opts.record;
+    var req;
+    var uploadDirname = opts.uploadDirname;
+
+    // XXX validate record
```
[the remark this answers is not in the document]
Comment (joshwilsdon) [reply lib/garbage-mpu-parts-cleaner.js:343]: Yeah, the code does do asserts on the fields we're using now. I can remove this. Thanks.

File: lib/garbage-mpu-parts-cleaner.js (hunk 1 above)
Comment (joshwilsdon) [reply lib/garbage-mpu-parts-cleaner.js:467]: I'll add a comment about that. And an assert.

**Pull request merged**